[Completion] Skip `tryOptimizeGenericDisjunction` if there's a completion child

hamishknight opened the pull request:
Attempting to favor a disjunction choice here can lead to unhelpful results if there's e.g a code completion token argument, since it acts as a placeholder.

rdar://127844278

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -6658,6 +6658,16 @@ static bool shouldHaveDirectCalleeOverload(const CallExpr *callExpr) {
 
+ASTNode ConstraintSystem::includingParentApply(ASTNode node) {
+  if (auto *expr = getAsExpr(node)) {
+    if (auto apply = dyn_cast_or_null<ApplyExpr>(getParentExpr(expr))) {
```
Nit: can replace `dyn_cast_or_null` with `getAsExpr<ApplyExpr>`

hamishknight committed:
[CS] NFC: Factor out `includingParentApply`

hamishknight committed:
[Completion] Skip `tryOptimizeGenericDisjunction` if there's a completion child

Attempting to favor a disjunction choice here
can lead to unhelpful results if there's e.g a code
completion token argument, since it acts as a
placeholder.

rdar://127844278

hamishknight commented:
@swift-ci please test

hamishknight commented:
Stress tester run: https://ci.swift.org/job/swift-PR-macos-with-sourcekit-stress-tester/358/

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
Stress tester result is UPASS
